### PR TITLE
Add QoS Options Support

### PR DESCRIPTION
**Pull request by yukikatase**
Add QoS Options Support
(https://ecl.ntt.com/en/documents/api-references/network/azure/qos_option.html#list-qos-option)

### CODE REVIEW

File: ecl/network/v2/qos_options/requests.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package qos_options
+
+import (
+	// "encoding/json"
```
Comment (ghost) [inline ecl/network/v2/qos_options/requests.go]: I would remove this line.

File: ecl/network/v2/qos_options/results.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// GetResult is the result of Get operations. Call its Extract method to
+// interpret it as a QoSOpt.
+type GetResult struct {
+	commonResult
+}
+
+// QoSOpt represents a QoS option.
+type QoSOpt struct {
```
Comment (ghost) [inline ecl/network/v2/qos_options/results.go]: Probably `QoSOption` is not too long, so I would modify struct name.

**Review by ghost — CHANGES_REQUESTED** [review]

File: ecl/network/v2/qos_options/results.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+// ExtractQoSOptions accepts a Page struct, specifically a QoSOptionPage struct,
+// and extracts the elements into a slice of ListOpts structs. In other words,
+// a generic collection is mapped into a relevant slice.
+func ExtractQoSOptions(r pagination.Page) ([]QoSOpt, error) {
+	var s []QoSOpt
+	err := ExtractQoSOptionsInto(r, &s)
+	return s, err
+}
+
+func ExtractQoSOptionsInto(r pagination.Page, v interface{}) error {
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/results.go]: I think you don't need to define this function(received interface{} as parameter) separately. Please move it into `ExtractQoSOptions`. The same applies to `ExtractInto`.

File: ecl/network/v2/qos_options/testing/fixtures.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+			  "interdc_service_id"	: "45678",
+			  "internet_service_id" : "910111213",
+			  "name"				: "100Gbps-BestEffort",
+			  "qos_type"			: "besteffort",
+			  "service_type"		: "gcp",
+			  "status"				: "ACTIVE",
+			  "vpn_service_id"		: "613b6f71-ec65-4d51-93f8-ed43eaf4fd15"
+		}
+	]
+}`
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/fixtures.go]:
```suggestion
}
`
```

File: ecl/network/v2/qos_options/testing/fixtures.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+const GetResponse = `{
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/fixtures.go]: Please insert blank line between variables.
```suggestion

const GetResponse = `
{
```

File: ecl/network/v2/qos_options/testing/fixtures.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		  "id"					: "2c649b8e-f007-4d90-b208-9b8710937a94",
+		  "interdc_service_id"	: "34567",
+		  "internet_service_id" : "89101112",
+		  "name"				: "100Gbps-Guaranteed",
+		  "qos_type"			: "guarantee",
+		  "service_type"		: "vpn",
+		  "status"				: "ACTIVE",
+		  "vpn_service_id"		: ""
+	}
+}`
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/fixtures.go]:
```suggestion
}
`
```

File: ecl/network/v2/qos_options/testing/fixtures.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
+package testing
+
+import "github.com/nttcom/eclcloud/ecl/network/v2/qos_options"
+
+const ListResponse = `
+{
+	"qos_options": [
+	  	{
+			  "aws_service_id"		: "3472d029-8378-424c-aaaf-150edc08947b",
+			  "azure_service_id"	: "",
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/fixtures.go]: API Doc does not show blank string but null. The same applies to the other IDs. Please confirm which is correct.
https://ecl.ntt.com/en/documents/api-references/network/azure/qos_option.html#sample-response-body

File: ecl/network/v2/qos_options/testing/fixtures.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
+package testing
+
+import "github.com/nttcom/eclcloud/ecl/network/v2/qos_options"
+
+const ListResponse = `
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/fixtures.go:5]: I think only one service ID(e.g. aws_service_id) is not blank and the other IDs are blank. I mean two or more IDs never are not blank at the same time. Can you confirm actual api response?

File: ecl/network/v2/qos_options/testing/request_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	})
+
+	client := fake.ServiceClient()
+	count := 0
+
+	qos_options.List(client, qos_options.ListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+		count++
+		actual, err := qos_options.ExtractQoSOptions(page)
+		if err != nil {
+			t.Errorf("Failed to extrace ports: %v", err)
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/request_test.go]: Typo
```suggestion
			t.Errorf("Failed to extract QoS options: %v", err)
```

File: ecl/network/v2/qos_options/testing/request_test.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	if count != 1 {
+		t.Errorf("Expected 1 page, got %d", count)
+	}
+}
+
+func TestGetQoS(t *testing.T) {
+	th.SetupHTTP()
+	defer th.TeardownHTTP()
+
+	th.Mux.HandleFunc("/v2.0/qos_options/2c649b8e-f007-4d90-b208-9b8710937a94", func(w http.ResponseWriter, r *http.Request) {
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/request_test.go]: Can you introduce the ID(2c649b8e-f007-4d90-b208-9b8710937a94) to variable?

**Review by kazyshr — CHANGES_REQUESTED** [review]
Great job! Please confirm my comments.

**Review by ghost — APPROVED** [review]
LGTM:)

File: ecl/network/v2/qos_options/testing/request_test.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	})
+
+	client := fake.ServiceClient()
+	count := 0
+
+	qos_options.List(client, qos_options.ListOpts{}).EachPage(func(page pagination.Page) (bool, error) {
+		count++
+		actual, err := qos_options.ExtractQoSOptions(page)
+		if err != nil {
+			t.Errorf("Failed to extrace extract QoS options: %v", err)
```
Comment (kazyshr) [inline ecl/network/v2/qos_options/testing/request_test.go]:
```suggestion
			t.Errorf("Failed to extract QoS options: %v", err)
```

**Review by kazyshr — APPROVED** [review]
Looks great! Thanks for the fixes.

**Pull request merged**